### DISCUSSION TITLE
Undocumented behaviour

### DISCUSSION THREAD

**Discussion by smith558** [nfstream/nfstream#174] [General]
I am wondering what the actual definition of the `bidirectional_first_seen_ms` flow feature is. I understand it represents "the timestamp (in milliseconds) when a bidirectional flow was first seen".

However, it is not clear to me what the behaviour is when the flow is unidirectional. Are unidirectional flows considered as a subset of bidirectional flows, and thus `bidirectional_first_seen_ms` would be equal to `src2dst_first_seen_ms`? Or are they not, and the value might be set to a default value (e.g., `None`, `-1`, or `0`)?

I feel like this is missing in the API documentation https://www.nfstream.org/docs/api.

**Comment by drnpkr** [comment]
Unidirectional flows and bidirectional flows are related in that a bidirectional flow is made up of two unidirectional flows that occur in opposite directions.

In a bidirectional flow, data is transmitted back and forth between two devices. Each device sends data to the other in a separate unidirectional flow, which is identified by its source and destination IP addresses and port numbers. For example, one device sends data in a unidirectional flow from its source IP address and port number to the destination IP address and port number of the other device, and the other device sends data in a separate unidirectional flow from its source IP address and port number to the first device's IP address and port number.

Both unidirectional flows make up the bidirectional flow, and they are typically treated as a single entity for analysis and classification purposes. In some cases, only one of the unidirectional flows may be present in the network traffic, but it can still be classified as part of a bidirectional flow based on its associated flow characteristics.

As per your question: Are unidirectional flows considered as a subset of bidirectional flows, and thus bidirectional_first_seen_ms would be equal to src2dst_first_seen_ms? Yes (but it can also be dst2src_first_seen_ms).

By the way, I answered your question in Gitter. Did not you find my answer satisfactory enough there?

Edit: To avoid any confusion, I want to clarify that with NFStream, in the above context, the bidirectional_first_seen_ms value will always be equal to the src2dst_first_seen_ms value. However, in theory, there are some scenarios where the flow direction may be reversed.

For example, in a 3-way handshake (SYN, SYN-ACK, ACK) where the first packet is lost, the flow may be initiated with the second packet (SYN-ACK) instead of the first packet (SYN). In this case, the src2dst_first_seen_ms value will still be used, even though the theoretical flow direction would be reversed.

**Reply by smith558** [reply]
Hello, and thank you! Just saw your Gitter message.
